Add generic DefaultHash<> for types with hash() function and equality operator

anttijk opened the pull request:
#### 106837a9c4ea12d4618f4fdb5d28f8a48b9abed3
<pre>
Add generic DefaultHash&lt;&gt; for types with hash() function and equality operator
<a href="https://bugs.webkit.org/show_bug.cgi?id=301803">https://bugs.webkit.org/show_bug.cgi?id=301803</a>
<a href="https://rdar.apple.com/163849870">rdar://163849870</a>

Reviewed by Darin Adler and Sam Weinig.

With this any moveable class or struct can be turned into a hash table key with minimal ceremony:

struct Foo {
    unsigned a;
    Field b;

    bool operator==(const Foo&amp;) = default;
    unsigned hash() const { return computeHash(a, b); }
};

HashSet&lt;GenericHashKey&lt;Foo&gt;&gt; set;

Also remove a bunch of now-unneeded DefaultHashes that match this pattern (with safeToCompareToEmptyOrDeleted = false).

* Source/JavaScriptCore/dfg/DFGIntegerCheckCombiningPhase.cpp:
(JSC::DFG::IntegerCheckCombiningPhase::RangeKey::Hash::hash): Deleted.
(JSC::DFG::IntegerCheckCombiningPhase::RangeKey::Hash::equal): Deleted.
* Source/JavaScriptCore/ftl/FTLSlowPathCallKey.h:
(JSC::FTL::SlowPathCallKeyHash::hash): Deleted.
(JSC::FTL::SlowPathCallKeyHash::equal): Deleted.
* Source/JavaScriptCore/runtime/FunctionHasExecutedCache.h:
(JSC::FunctionHasExecutedCache::FunctionRange::Hash::hash): Deleted.
(JSC::FunctionHasExecutedCache::FunctionRange::Hash::equal): Deleted.
* Source/JavaScriptCore/runtime/TemplateObjectDescriptor.h:
(JSC::TemplateObjectDescriptor::Hasher::hash): Deleted.
(JSC::TemplateObjectDescriptor::Hasher::equal): Deleted.
* Source/JavaScriptCore/runtime/TypeLocationCache.h:
(JSC::TypeLocationCache::LocationKey::Hash::hash): Deleted.
(JSC::TypeLocationCache::LocationKey::Hash::equal): Deleted.
* Source/WTF/wtf/HashFunctions.h:
(WTF::requires):

Add HashableWithMemberFunction concept.

(WTF::MemberBasedHash::hash):
(WTF::MemberBasedHash::equal):

Use it.

* Source/WebCore/style/RuleFeature.h:
(WebCore::Style::RuleFeatureDeduplicationKey::Hash::hash): Deleted.
(WebCore::Style::RuleFeatureDeduplicationKey::Hash::equal): Deleted.
* Source/WebKit/Shared/WebPageNetworkParameters.h:
(WTF::DefaultHash&lt;WebKit::WebPageNetworkParameters&gt;::hash): Deleted.
(WTF::DefaultHash&lt;WebKit::WebPageNetworkParameters&gt;::equal): Deleted.

Canonical link: <a href="https://commits.webkit.org/302440@main">https://commits.webkit.org/302440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5886110ae6e3bb9bb7034acd16c932a2c350c509

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129098 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1357 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39934 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136477 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80472 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4e15695c-3dd7-4dae-8102-62987be00fc7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130969 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1234 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/98294 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66168 "") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b5086d28-9416-4855-8f02-bf15692a97f3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132045 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/998 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115645 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78939 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dc6098b3-211c-4658-8bcd-c526025eb9ce) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/922 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33761 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79756 "Built successfully") | | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/121090 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109369 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34258 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138950 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/127550 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1150 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/1118 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/106830 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1202 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111981 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106657 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27151 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/944 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30505 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53688 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1223 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64575 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/160564 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1049 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40075 "Found 5 new JSC stress test failures: mozilla-tests.yaml/ecma/Date/15.9.5.14.js.mozilla, mozilla-tests.yaml/ecma/Date/15.9.5.14.js.mozilla-baseline, mozilla-tests.yaml/ecma/Date/15.9.5.14.js.mozilla-dfg-eager-no-cjit-validate-phases, mozilla-tests.yaml/ecma/Date/15.9.5.14.js.mozilla-llint, mozilla-tests.yaml/ecma/Date/15.9.5.14.js.mozilla-no-ftl (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1095 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1147 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/HashFunctions.h:
```diff
@@ -259,6 +259,17 @@ namespace WTF {
+        { t.hash() } -> std::same_as<unsigned>;
+        { t == t } -> std::same_as<bool>;
+    };
+    template<typename T> requires HashableType<T> struct DefaultHash<T>  {
```
Is this another way to write the same thing?
```suggestion
    template<HashableType T> struct DefaultHash<T> [
```
Would it be useful to have this hash under a different name so you could specialize with it as a base? Maybe `MemberBasedHash`?

darinadler reviewed (approved):
Maybe we could similarly allow `safeToCompareEmptyOrDeleted` to be a data or function member of the type and reduce the need to specialize the hash even further?

weinig commented on Source/WTF/wtf/HashFunctions.h:
```diff
@@ -259,6 +259,18 @@ namespace WTF {
     template<typename... Types> struct DefaultHash<std::tuple<Types...>> : TupleHash<Types...> { };
 
+    // Default hash for any type with a hash() member function and an equality operator.
+    template<typename T> concept HashableWithMemberFunction = requires(const T& t) {
```
You may be able to write this as:

```c++
    template<typename T> concept HashableWithMemberFunction = std::equality_comparable<T> && requires(const T& t) {
        { t.hash() } -> std::same_as<unsigned>;
    };
```

weinig reviewed (approved):
Really nice simplification!

anttijk committed:
Add generic DefaultHash<> for types with hash() function and equality operator
https://bugs.webkit.org/show_bug.cgi?id=301803
rdar://163849870

Reviewed by Darin Adler and Sam Weinig.

With this any moveable class or struct can be turned into a hash table key with minimal ceremony:

struct Foo {
    unsigned a;
    Field b;

    bool operator==(const Foo&) = default;
    unsigned hash() const { return computeHash(a, b); }
};

HashSet<GenericHashKey<Foo>> set;

Also remove a bunch of now-unneeded DefaultHashes that match this pattern (with safeToCompareToEmptyOrDeleted = false).

* Source/JavaScriptCore/dfg/DFGIntegerCheckCombiningPhase.cpp:
(JSC::DFG::IntegerCheckCombiningPhase::RangeKey::Hash::hash): Deleted.
(JSC::DFG::IntegerCheckCombiningPhase::RangeKey::Hash::equal): Deleted.
* Source/JavaScriptCore/ftl/FTLSlowPathCallKey.h:
(JSC::FTL::SlowPathCallKeyHash::hash): Deleted.
(JSC::FTL::SlowPathCallKeyHash::equal): Deleted.
* Source/JavaScriptCore/runtime/FunctionHasExecutedCache.h:
(JSC::FunctionHasExecutedCache::FunctionRange::Hash::hash): Deleted.
(JSC::FunctionHasExecutedCache::FunctionRange::Hash::equal): Deleted.
* Source/JavaScriptCore/runtime/TemplateObjectDescriptor.h:
(JSC::TemplateObjectDescriptor::Hasher::hash): Deleted.
(JSC::TemplateObjectDescriptor::Hasher::equal): Deleted.
* Source/JavaScriptCore/runtime/TypeLocationCache.h:
(JSC::TypeLocationCache::LocationKey::Hash::hash): Deleted.
(JSC::TypeLocationCache::LocationKey::Hash::equal): Deleted.
* Source/WTF/wtf/HashFunctions.h:
(WTF::requires):

Add HashableWithMemberFunction concept.

(WTF::MemberBasedHash::hash):
(WTF::MemberBasedHash::equal):

Use it.

* Source/WebCore/style/RuleFeature.h:
(WebCore::Style::RuleFeatureDeduplicationKey::Hash::hash): Deleted.
(WebCore::Style::RuleFeatureDeduplicationKey::Hash::equal): Deleted.
* Source/WebKit/Shared/WebPageNetworkParameters.h:
(WTF::DefaultHash<WebKit::WebPageNetworkParameters>::hash): Deleted.
(WTF::DefaultHash<WebKit::WebPageNetworkParameters>::equal): Deleted.

Canonical link: https://commits.webkit.org/302440@main
